### PR TITLE
Enable setting environment variables via Hooks

### PR DESCRIPTION
**Pull request by AlfredBroda**
This allows for setting application environment via Hooks (for remote configuration retrieval purposes).

### CODE REVIEW

**Comment by medzin** [comment]
Why I see "Fix Travis configuration for releses #22" in this PR?

**Comment by AlfredBroda** [comment]
Fixed all the conflicts... for now.

File: command.go [hunk 1]
```diff
@@ -121,7 +122,7 @@ func (c *cancellableCommand) Stop(gracePeriod time.Duration) {
 }
 
 // NewCommand returns a new command based on passed CommandInfo.
-func NewCommand(commandInfo mesos.CommandInfo, env []string, options ...func(*exec.Cmd) error) (Command, error) {
+func NewCommand(commandInfo mesos.CommandInfo, env hook.Env, options ...func(*exec.Cmd) error) (Command, error) {
```
Comment (janisz) [inline command.go]: `Env` should not be defined in `hook` package

File: hook/consul/hook.go [hunk 2]
```diff
@@ -46,15 +46,15 @@ type Config struct {
...
 	case hook.AfterTaskHealthyEvent:
-		return h.RegisterIntoConsul(event.TaskInfo)
+		return hook.Env{}, h.RegisterIntoConsul(event.TaskInfo)
 	case hook.BeforeTerminateEvent:
-		return h.DeregisterFromConsul(event.TaskInfo)
+		return hook.Env{}, h.DeregisterFromConsul(event.TaskInfo)
 	default:
 		log.Debugf("Received unsupported event type %s - ignoring", event.Type)
-		return nil // ignore unsupported events
+		return hook.Env{}, nil // ignore unsupported events
```
Comment (janisz) [inline hook/consul/hook.go]: IMO `nil` will be better than `hook.Env{}`

Comment (medzin) [reply hook/consul/hook.go]: It's not a problem really - you have to perform error check before you will use that value.

Comment (AlfredBroda) [reply hook/consul/hook.go]: And deal with mocking in tests. Will try that approach (again).

Comment (medzin) [reply hook/consul/hook.go]: I mean, it does not make sense to return pointer here.

File: command.go (hunk 1 above)
Comment (AlfredBroda) [reply command.go]: Where do you suggest to put it? executor.go? command.go?

File: executor.go [hunk 3]
```diff
@@ -343,7 +343,16 @@ func (e *Executor) launchTask(taskInfo mesos.TaskInfo) (Command, error) {
 		cmdOption = ForwardCmdOutput()
 	}
 
-	cmd, err := NewCommand(commandInfo, env, cmdOption)
+	beforeStartEvent := hook.Event{
+		Type:     hook.BeforeTaskStartEvent,
+		TaskInfo: mesosutils.TaskInfo{TaskInfo: taskInfo},
+	}
+	additionalEnv, err := e.hookManager.HandleEvent(beforeStartEvent, false)
```
Comment (medzin) [inline executor.go]: You have to merge `env` and `additionalEnv` slices or services will not have access to certificate or `MESOS_` variables.

Comment (medzin) [reply executor.go]: Also add test case for this - the current ones did not catch this problem

File: command.go (hunk 1 above)
Comment (AlfredBroda) [reply command.go]: Reverted back to []string

File: hook/consul/hook.go (hunk 2 above)
Comment (AlfredBroda) [reply hook/consul/hook.go]: Returning nil then.

**Comment by medzin** [comment]
@AlfredBroda explain in commit message why we are enabling this.

File: hook/hook.go [hunk 4]
```diff
@@ -25,12 +25,15 @@ type Event struct {
 	TaskInfo mesosutils.TaskInfo
 }
 
+// Env is a container type for environment variables.
```
Comment (medzin) [inline hook/hook.go]: Add info about format. Most people will expect map here - not slice.

File: mesosutils/taskinfo_test.go [hunk 5]
```diff
@@ -263,3 +263,23 @@ func TestGetWeightReturnsWeightFromTagLabel(t *testing.T) {
...
+			Environment: &mesos.Environment{Variables: vars},
+		},
+	}
+	taskInfo := TaskInfo{
+		TaskInfo: mesosTaskInfo,
+	}
+
+	returnedValue := taskInfo.FindEnvValue(key)
+
+	require.Equal(t, returnedValue, expectedValue)
```
Comment (medzin) [inline mesosutils/taskinfo_test.go]: Signature of `Equal` is:

`func Equal(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})`

(switch `returnedValue` and `expectedValue`)

File: executor.go [hunk 6]
```diff
@@ -405,8 +414,8 @@ func (e *Executor) shutDown(taskInfo mesos.TaskInfo, cmd Command) {
 		Type:     hook.BeforeTerminateEvent,
 		TaskInfo: mesosutils.TaskInfo{TaskInfo: taskInfo},
 	}
-	_ = e.hookManager.HandleEvent(beforeTerminateEvent, true) // ignore errors here, so every hook will have a chance to be called
-	cmd.Stop(gracePeriod)                                     // blocking call
+	_, _ = e.hookManager.HandleEvent(beforeTerminateEvent, true) // ignore errors here, so every hook will have a chance to be called
```
Comment (janisz) [inline executor.go]: I think we should log errors

File: hook/consul/hook.go [hunk 7]
```diff
@@ -46,15 +46,15 @@ type Config struct {
 
 // HandleEvent calls appropriate hook functions that correspond to supported
 // event types. Unsupported events are ignored.
-func (h *Hook) HandleEvent(event hook.Event) error {
+func (h *Hook) HandleEvent(event hook.Event) (hook.Env, error) {
```
Comment (janisz) [inline hook/consul/hook.go]: Please document when HandleEvent returns `hook.Env`

File: hook/hook.go [hunk 8]
```diff
@@ -25,12 +25,15 @@ type Event struct {
 	TaskInfo mesosutils.TaskInfo
 }
 
+// Env is a container for os.Environ style list of combined environment variable strings.
+type Env []string
```
Comment (janisz) [inline hook/hook.go]: I think this should not be in this package

File: hook/hook.go [hunk 9]
```diff
@@ -25,12 +25,15 @@ type Event struct {
...
+
 // Hook is an interface for various executor extensions, that can add some actions
 // during task lifecycle events.
 type Hook interface {
 	// HandleEvent is called when any of defined executor task events occurs.
 	// Received events may be handled in any way, but hook should ignore unknown or
 	// unsupported ones. Call to this function will block executor process until
 	// it returns. Order of received event types is undefined.
-	HandleEvent(Event) error
+	HandleEvent(Event) (Env, error)
```
Comment (janisz) [inline hook/hook.go]: Update the godoc

File: hook/hook.go (hunk 8 above)
Comment (medzin) [reply hook/hook.go]: Suggest another then 😄

**Pull request merged**